lib/wallet/src/Cardano/Wallet/DB.hs

```diff
@@ -332,6 +343,308 @@ data DBLayer m s k = forall stm. (MonadIO stm, MonadFail stm) => DBLayer
+        -- to the newest.
+    }
+
+-- | A database layer for storing 'WalletMetadata'.
+data DBWalletMeta stm = DBWalletMeta
+    { putWalletMeta_
+        :: WalletId
+        -> WalletMetadata
+        -> ExceptT ErrNoSuchWallet stm ()
+        -- ^ Replace an existing wallet metadata with the given one.
+        --
+        -- If the wallet doesn't exist, this operation returns an error
+
+    , readWalletMeta_
+        :: WalletId
+        -> stm (Maybe WalletMetadata)
+        -- ^ Fetch a wallet metadata, if they exist.
+        --
+        -- Return 'Nothing' if there's no such wallet.
+    }
+
+-- | A database layer for storing delegation certificates
+-- and the reward balance.
+data DBDelegation stm = DBDelegation
+    { isStakeKeyRegistered_
+        :: WalletId
+        -> ExceptT ErrNoSuchWallet stm Bool
+
+    , putDelegationCertificate_
+        :: WalletId
+        -> DelegationCertificate
+        -> SlotNo
+        -> ExceptT ErrNoSuchWallet stm ()
+        -- ^ Binds a stake pool id to a wallet. This will have an influence on
+        -- the wallet metadata: the last known certificate will indicate to
+        -- which pool a wallet is currently delegating.
+        --
+        -- This is done separately from 'putWalletMeta' because certificate
+        -- declarations are:
+        --
+        -- 1. Stored on-chain.
+        -- 2. Affected by rollbacks (or said differently, tied to a 'SlotNo').
+
+    , putDelegationRewardBalance_
+        :: WalletId
+        -> Coin
+        -> ExceptT ErrNoSuchWallet stm ()
+        -- ^ Store the latest known reward account balance.
+        --
+        -- This is separate from checkpoints because the data corresponds to the
+        -- node tip.
+        -- This is separate from putWalletMeta because it's not meta data
+
+    , readDelegationRewardBalance_
+        :: WalletId
+        -> stm Coin
+        -- ^ Get the reward account balance.
+        --
+        -- Returns zero if the wallet isn't found or if wallet hasn't delegated
+        -- stake.
+    }
+
+-- | A database layer that stores the transaction history.
+data DBTxHistory stm = DBTxHistory
+    { putTxHistory_
+        :: WalletId
+        -> [(Tx, TxMeta)]
+        -> ExceptT ErrNoSuchWallet stm ()
+        -- ^ Augments the transaction history for a known wallet.
+        --
+        -- If an entry for a particular transaction already exists it is not
+        -- altered nor merged (just ignored).
+        --
+        -- If the wallet doesn't exist, this operation returns an error.
+
+    , readTxHistory_
+        :: WalletId
+        -> Maybe Coin
+        -> SortOrder
+        -> Range SlotNo
+        -> Maybe TxStatus
+        -> stm [TransactionInfo]
+        -- ^ Fetch the current transaction history of a known wallet, ordered by
+        -- descending slot number.
+        --
+        -- Returns an empty list if the wallet isn't found.
```

Comment:
```
        -- Returns an empty list if the wallet isn't found.
```
This is inconsistent with how adjacent functions error out on missing wallets. Is there a reason not to `ExceptT ErrNoSuchWallet` ?

Reply:
Yes — "it was appropriate before", and for the next pull requests, I plan to kill most of the `ExceptT ErrNoSuchWallet` noise.